### PR TITLE
Extract NixOS options documentation generation to a function

### PR DESCRIPTION
**Pull request by domenkozar**
Motivation is to support other repositories containing nixos
modules that would like to generate options documentation:

- nix-darwin
- private repos
- arion
- ??
 
Refs https://github.com/hercules-ci/arion/issues/10

### CODE REVIEW

File: nixos/lib/make-options-doc/default.nix [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  optionsListVisible = lib.filter (opt: opt.visible && !opt.internal) (lib.optionAttrSetToDocList options);
+
+  # Customly sort option list for the man page.
+  optionsList = lib.sort optionLess optionsListDesc;
+
+  # Convert the list of options into an XML file.
+  optionsXML = builtins.toFile "options.xml" (builtins.toXML optionsList);
+
+in rec {
+  # The NixOS options in JSON format.
```
Comment (tazjin) [inline nixos/lib/make-options-doc/default.nix]: Could we get an `optionsNix` for other Nix-internal use-cases? (to avoid reading the JSON back in)

Comment (domenkozar) [reply nixos/lib/make-options-doc/default.nix]: I plan to add that and asciidoc variation, but I'd like to do that as a separate PR to limit the scope here.

**Comment by rycee** [comment]
Very nice to have some cleanups! I have very slowly been working on something similar for use in Home Manager. Basically I've been wanting to extract the whole documentation generation framework into a project of its own, I uploaded the latest code I have here: [nmd](https://gitlab.com/rycee/nmd).

It contains some cleaned up code for [option documentation](https://gitlab.com/rycee/nmd/blob/master/lib/modules-doc.nix) although I can't recall whether it actually works or not. But it is more general than the one in both NixOS and Home Manager in that, for example, the URL for the module is generated by a given function argument rather than being hard coded in the xsl file.

**Comment by domenkozar** [comment]
@rycee that's really cool :) I'd like to improve this interface further until 19.09 is out, so we can incorporate some of your ideas.

File: nixos/lib/make-options-doc/default.nix (hunk 1 above)
Comment (domenkozar) [reply nixos/lib/make-options-doc/default.nix]: https://github.com/NixOS/nixpkgs/pull/66436

**Comment by rycee** [comment]
@domenkozar FWIW I played around a bit more with nmd and now Home Manager uses it to build its documentation: https://github.com/rycee/home-manager/blob/master/doc/default.nix. It's a bit ugly but at least it produces the same output as before.

**Pull request merged**